framework/Optimizers/GradientBasedOptimizer.py

```diff
@@ -265,86 +267,124 @@ def checkConvergence(self):
-        ## convergence values
-        # gradient norm
-        gradNorm  = self.counter['gradNormHistory'][traj][0]
+    # if trajectory is already converged, no need to update. #FIXME then why was this method called?
+    if self.convergeTraj[traj]:
+      return
+
+    # first, check if we're at varsUpdate 0 (first entry); if so, we are at our first point
+    if varsUpdate == 0:
+      # we don't have enough points to decide to accept or reject the new point, so accept it as the initial point
+      self.raiseADebug('Accepting first point, since we have no rejection criteria.')
+      self.status[traj]['reason'] = 'found new opt point'
+      return
+
+    #otherwise, we need to accept/reject point and check convergence
+    currentInputDenorm = self.denormalizeData(self.optVarsHist[traj][self.counter['varsUpdate'][traj]])
+
+    ## first, determine if we want to keep the new point
+    # obtain the old loss value
+    oldLossVal = self.counter['recentOptHist'][traj][0]['output']
+    # see if new point is better than old point
+    newerIsBetter = self.checkIfBetter(currentLossVal,oldLossVal)
+    # if this was a recommended preconditioning point, we should not be converged.
+    pointFromRecommendation = self.status[traj]['reason'] == 'received recommended point'
+    # if improved, keep it and move forward; otherwise, reject it and recommend cutting step size
+    if newerIsBetter:
+      self.status[traj]['reason'] = 'found new opt point'
+      self.raiseADebug('Accepting potential opt point for improved loss value')
+      #TODO REWORK this belongs in the base class optimizer; grad shouldn't know about multilevel!!
+      #  -> this parameter is how multilevel knows that a successful perturbation of an outer loop has been performed
+      #  maybe implement a "acceptPoint" method in base class?
+      self.mlActiveSpaceSteps[traj] += 1
+    else:
+      self.status[traj]['reason'] = 'rejecting bad opt point'
+      self.raiseADebug('Rejecting potential opt point for worse loss value. old: "{}", new: "{}"'.format(oldLossVal,currentLossVal))
+      # cut the next step size to hopefully stay in the valley instead of climb up the other side
+      self.recommendToGain[traj] = 'cut'
+
+    ## determine convergence
+    if pointFromRecommendation:
+      self.raiseAMessage('Setting convergence for Trajectory "{}" to "False" because of preconditioning.'.format(traj))
+      converged = False
+    else:
+      self.raiseAMessage('Checking convergence for Trajectory "{}":'.format(traj))
+      self.convergenceProgress[traj] = {} # tracks progress for grad norm, abs, rel tolerances
+      converged = False                   # updated for each individual criterion using "or" (pass one, pass all)
+      #printing utility
+      printString = '    {:<21}: {:<5}'
+      printVals = printString + ' (check: {:>+9.2e} < {:>+9.2e}, diff: {:>9.2e})'
+      def printProgress(name,boolCheck,test,gold):
```

Comment:
I think if we can come up with a use of this somewhere else, then it might be worth moving to utils.  Otherwise, I think a one line function can stay local.